pkg/providers/cloudstack/validator.go

```diff
@@ -0,0 +1,192 @@
+package cloudstack
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"strconv"
+	"strings"
+
+	anywherev1 "github.com/aws/eks-anywhere/pkg/api/v1alpha1"
+	"github.com/aws/eks-anywhere/pkg/logger"
+)
+
+// TODO: Move this to cloudstack.go provider file
+const controlEndpointDefaultPort = "6443"
+
+type Validator struct {
+	cmk ProviderCmkClient
+}
+
+func NewValidator(cmk ProviderCmkClient) *Validator {
+	return &Validator{
+		cmk: cmk,
+	}
+}
+
+type ProviderCmkClient interface {
+	ValidateCloudStackConnection(ctx context.Context) error
+	ValidateServiceOfferingPresent(ctx context.Context, domain string, zone anywherev1.CloudStackResourceRef, account string, serviceOffering anywherev1.CloudStackResourceRef) error
+	ValidateTemplatePresent(ctx context.Context, domain string, zone anywherev1.CloudStackResourceRef, account string, template anywherev1.CloudStackResourceRef) error
+	ValidateAffinityGroupsPresent(ctx context.Context, domain string, zone anywherev1.CloudStackResourceRef, account string, affinityGroupIds []string) error
+	ValidateZonePresent(ctx context.Context, zone anywherev1.CloudStackResourceRef) error
+}
+
+func (v *Validator) validateCloudStackAccess(ctx context.Context) error {
+	if err := v.cmk.ValidateCloudStackConnection(ctx); err != nil {
+		return fmt.Errorf("failed validating connection to vCenter: %v", err)
+	}
+	logger.MarkPass("Connected to server")
+
+	return nil
+}
+
+func (v *Validator) ValidateCloudStackDatacenterConfig(ctx context.Context, datacenterConfig *anywherev1.CloudStackDatacenterConfig) error {
+	if err := v.cmk.ValidateZonePresent(ctx, datacenterConfig.Spec.Zone); err != nil {
+		return err
+	}
+	logger.MarkPass("Datacenter validated")
+
+	return nil
+}
+
+// TODO: dry out machine configs validations
+func (v *Validator) ValidateClusterMachineConfigs(ctx context.Context, cloudStackClusterSpec *spec) error {
+	var etcdMachineConfig *anywherev1.CloudStackMachineConfig
+
+	if len(cloudStackClusterSpec.Cluster.Spec.ControlPlaneConfiguration.Endpoint.Host) <= 0 {
+		return fmt.Errorf("cluster controlPlaneConfiguration.Endpoint.Host is not set or is empty")
+	}
+	if len(cloudStackClusterSpec.datacenterConfig.Spec.Domain) <= 0 {
+		return fmt.Errorf("CloudStackDatacenterConfig domain is not set or is empty")
+	}
+	if len(cloudStackClusterSpec.datacenterConfig.Spec.Zone.Value) <= 0 {
+		return fmt.Errorf("CloudStackDatacenterConfig zone is not set or is empty")
+	}
+	if len(cloudStackClusterSpec.datacenterConfig.Spec.Network.Value) <= 0 {
+		return fmt.Errorf("CloudStackDatacenterConfig network is not set or is empty")
+	}
+	if cloudStackClusterSpec.Cluster.Spec.ControlPlaneConfiguration.MachineGroupRef == nil {
+		return fmt.Errorf("must specify machineGroupRef for control plane")
+	}
+
+	if cloudStackClusterSpec.Cluster.Spec.WorkerNodeGroupConfigurations[0].MachineGroupRef == nil {
+		return fmt.Errorf("must specify machineGroupRef for worker nodes")
+	}
```

Comment:
Should this be a loop as well?
The one you have for
`for _, workerNodeGroupConfiguration := range cloudStackClusterSpec.Cluster.Spec.WorkerNodeGroupConfigurations {`

Reply:
Yes! Good catch.